pgml-docs/docs/blog/scaling-postgresml-to-one-million-requests-per-second.md

```diff
@@ -0,0 +1,252 @@
+
+By combining PgCat with PostgresML, we are aiming to build the next generation of machine learning infrastructure that can power anything from two-person startups, like us, to unicorns and massive enterprises, without the data ever leaving our favorite database.
+
+
+## Methodology
+
+### ML
+
+This time, we used an XGBoost model with 100 trees
+
+```postgresql
+SELECT * FROM pgml.train(
+	'flights',
+	task => 'regression',
+	relation_name => 'flights_mat_3',
+	y_column_name => 'depdelayminutes',
+	algorithm => 'xgboost',
+	hyperparams => '{"n_estimators": 100 }',
+	runtime => 'rust'
+);
+```
+
+and fetched our predictions the usual way
+
+```postgresql
+SELECT pgml.predict(
+	'flights',
+	ARRAY[
+		year,
+		quarter,
+		month,
+		distance,
+		dayofweek,
+		dayofmonth,
+		flight_number_operating_airline,
+		originairportid,
+		destairportid,
+		flight_number_marketing_airline,
+		departure
+	]
+) AS prediction
+FROM flights_mat_3 LIMIT :limit;
+```
+
+where `:limit` is the batch size of 1, 5, and 20, depending on the benchmark.
+
+#### Model
+
+The model is roughly the same as the one we used in our previous [post](/blog/postgresml-is-8x-faster-than-python-http-microservices), with just one extra feature added, which improved R<sup>2</sup> a little bit.
+
+### Hardware
+
+#### Client
+The client was a `c5n.4xlarge` box on EC2. We chose the `c5n` class to have the 100 GBit NIC, since we wanted it to saturate our network as much as possible. Thousands of clients were simulated using [`pgbench`](https://www.postgresql.org/docs/current/pgbench.html) with `-c n` where `n` is the number of clients.
+
+#### PgCat Pooler
+PgCat, written in asynchronous Rust, was running on `c5.xlarge` machines (4 vCPUs, 8GB RAM) with 4 Tokio workers. We used between 1 and 35 machines, and scaled them in increments of 5-20 at a time.
+
+The pooler did a decent amount of work around parsing queries, making sure they are read-only `SELECT`s, and routing them, at random, to replicas. If any replica was down for any reason, it would route around it to remaining machines.
+
+#### Postgres Replicas
+Postgres replicas were running on `c5.9xlarge` machines with 36 vCPUs and 72 GB of RAM. The hot dataset fits entirely in memory. The servers were intentionally saturated to maximum capacity before scaling up to test queuing and graceful degradation of performance.
+
+
+#### Raw Results
+
+Raw latency data is available [here](https://static.postgresml.org/benchmarks/reads-latency.csv) and raw throughput data is available [here](https://static.postgresml.org/benchmarks/reads-throughput.csv).
+
+## Feedback
+
+Many thanks and ❤️ to all those who are supporting this endeavor. We’d love to hear feedback from the broader ML and Engineering community about applications and other real world scenarios to help prioritize our work. You can show your support by starring us on our [Github](https://github.com/postgresml/postgresml/).
+
+
+## We're Hiring!
```

Comment:
You only get 1 call to action. Otherwise people typically do nothing due to the paradox of choice.